### PR TITLE
Downclosure with iterated views on Z3

### PR DESCRIPTION
**Pull request by MattWindsor91**
This adds to #64 by putting in downclosure checks for Z3 **only**.  HSF and µZ3 are still to do.  There are also some changes to the pretty-printer to allow for properly colourised error messages.

Not sure whether it'd be best to merge #64 first, or merge this and close that one.

### CODE REVIEW

File: Reifier.fs [hunk 1]
```diff
@@ -4,62 +4,635 @@
...
+    let lookupFunc (protos : FuncDefiner<ProtoInfo>) (func : Func<_>)
+      : Result<ProtoInfo option, Error> =
+        // TODO(CaptainHayashi): proper doc comment
+        // TODO(CaptainHayashi): merge with Modeller.lookupFunc?
+        protos
+        |> Core.Instantiate.lookup func
+        |> mapMessages (curry LookupError func.Name)
+        |> lift (Option.map snd)
+
+    let downclosureMap (f : Var -> IntExpr<Sym<Var>>)
```
Comment (septract) [inline Reifier.fs]: need a doc comment

File: Reifier.fs [hunk 2]
```diff
@@ -4,62 +4,635 @@
...
+        (* Base downclosure for a view V[n](x):
+             D(emp) => D(V[0](x))
+           That is, the definition of V when the iterator is 0 can be no
+           stricter than the definition of emp.
+
+           The definition of emp can only mention global variables, so it does
+           not need to be freshened. *)
+        let check = mkImplies empDefn baseDefn
+
+        // Expression equivalence cannot handle symbols, so try remove them.
```
Comment (septract) [inline Reifier.fs]: doesn't approx, just fails if there are symbols in here.

Comment (septract) [reply Reifier.fs]: (this might be necessary if we have iteration and grasshopper)

File: Reifier.fs [hunk 3]
```diff
@@ -4,62 +4,635 @@
...
+        bind
+            (fun chk ->
+                // Pose that the check is a tautology, and use Z3 to check it.
+                let closed = equivHolds id (equiv chk BTrue)
+
+                if closed
+                then ok (func, defn)
+                else fail (BaseDownclosureError ([func], defn, empDefn)))
+            (check
+             |> Mapper.mapBoolCtx (tsfRemoveSym SymInIteratedConstraint) NoCtx
```
Comment (septract) [inline Reifier.fs]: 'removesym' - bit misleading: it fails if it finds a symbol...

Comment (septract) [reply Reifier.fs]: also: 'tsf'?

File: Reifier.fs [hunk 4]
```diff
@@ -4,62 +4,635 @@
...
+      (defn : BoolExpr<Sym<Var>>)
+      : Result<IteratedDFunc * BoolExpr<Sym<Var>>, Error> =
+        (* To do the inductive downclosure, we need to replace all instances of
+           the iterator in the definition with (iterator + 1) in one version. *)
+        let increment v =
+            if v = iterator
+            then mkAdd2 (AVar (Reg v)) (AInt 1L)
+            else AVar (Reg v)
+
+        let succDefn = downclosureMap increment defn
```
Comment (septract) [inline Reifier.fs]: 'substituteIterator' name?

File: Reifier.fs [hunk 5]
```diff
@@ -4,62 +4,635 @@
...
+        (* Base downclosure for a view V[n](x):
+             (0 <= n) => (D(V[n+1](x)) => D(V[n](x)))
+           That is, the definition of V when the iterator is n+1 implies the
+           definition of V when the iterator is n, for all positive n. *)
+        let check =
+            mkImplies
+                (mkLe (AInt 0L) (AVar (Reg iterator)))
+                (mkImplies succDefn defn)
+
+        // Expression equivalence cannot handle symbols, so try remove them.
```
Comment (septract) [inline Reifier.fs]: duplication?

File: Reifier.fs [hunk 6]
```diff
@@ -4,62 +4,635 @@
...
+        let empDefn =
+            definer
+            |> ViewDefiner.toSeq
+            |> Seq.tryFind (fst >> List.isEmpty)
+            |> Option.bind snd
+
+        let checkEach def definer =
+            ok def >>= checkDef empDefn vprotos >>= (fun _ -> ok definer)
+
+        definer |> ViewDefiner.toSeq |> seqBind checkEach definer
```
Comment (septract) [inline Reifier.fs]: Don't need to pass definer if it's in scope

**Comment by septract** [comment]
Is this ready to merge now?

**Comment by MattWindsor91** [comment]
Should be.  :shipit:

**Pull request merged**